File: packages/cli/src/utils/mcp-config.ts
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import chalk from 'chalk';

/**
 * MCP configuration types
 */
export interface McpServerConfig {
  command: string;
  args: string[];
  env?: Record<string, string>;
}

export interface McpConfigFile {
  mcpServers: Record<string, McpServerConfig>;
}

/**
 * MCP-capable AI tool configuration
 */
export interface McpToolConfig {
  id: string;
  name: string;
  description: string;
  configPath: string;
  configLocation: 'workspace' | 'global';
  usesProjectVariable: boolean; // Whether the tool supports ${PWD} or similar
  detection?: {
    directories?: string[];
    files?: string[];
  };
}

export type McpToolKey = 'claude' | 'vscode' | 'cursor' | 'windsurf';

/**
 * MCP tool configurations
 */
export const MCP_TOOL_CONFIGS: Record<McpToolKey, McpToolConfig> = {
  'claude': {
    id: 'claude',
    name: 'Claude',
    description: 'Claude (.mcp.json)',
    configPath: '.mcp.json',
    configLocation: 'workspace',
    usesProjectVariable: true,
    detection: {
      directories: ['.claude'],
      files: ['CLAUDE.md', '.claude.json'],
    },
  },
  vscode: {
    id: 'vscode',
    name: 'VS Code',
    description: 'VS Code with GitHub Copilot (.vscode/mcp.json)',
    configPath: '.vscode/mcp.json',
    configLocation: 'workspace',
    usesProjectVariable: false,
    detection: {
      directories: ['.vscode'],
    },
  },
  cursor: {
    id: 'cursor',
    name: 'Cursor',
    description: 'Cursor (.cursor/mcp.json)',
    configPath: '.cursor/mcp.json',
    configLocation: 'workspace',
    usesProjectVariable: false,
    detection: {
      directories: ['.cursor'],
      files: ['.cursorrules'],
    },
  },
  windsurf: {
    id: 'windsurf',
    name: 'Windsurf',
    description: 'Windsurf (.windsurf/mcp.json)',
    configPath: '.windsurf/mcp.json',
    configLocation: 'workspace',
    usesProjectVariable: false,
    detection: {
      directories: ['.windsurf'],
      files: ['.windsurfrules'],
    },
  },
};

/**
 * Detection result for MCP tools
 */
export interface McpToolDetectionResult {
  tool: McpToolKey;
  detected: boolean;
  reasons: string[];
}

/**
 * Detect which MCP-capable tools are likely in use
 */
export async function detectMcpTools(cwd: string): Promise<McpToolDetectionResult[]> {
  const results: McpToolDetectionResult[] = [];

  for (const [key, config] of Object.entries(MCP_TOOL_CONFIGS)) {
    const reasons: string[] = [];
    const detection = config.detection;

    if (!detection) {
      results.push({ tool: key as McpToolKey, detected: false, reasons: [] });
      continue;
    }

    // Check directories
    if (detection.directories) {
      for (const dir of detection.directories) {
        const dirPath = path.join(cwd, dir);
        try {
          const stat = await fs.stat(dirPath);
          if (stat.isDirectory()) {
            reasons.push(`${dir}/ directory found`);
          }
        } catch {
          // Directory doesn't exist
        }
      }
    }

    // Check files
    if (detection.files) {
      for (const file of detection.files) {
        const filePath = path.join(cwd, file);
        try {
          await fs.access(filePath);
          reasons.push(`${file} found`);
        } catch {
          // File doesn't exist
        }
      }
    }

    results.push({
      tool: key as McpToolKey,
      detected: reasons.length > 0,
      reasons,
    });
  }

  return results;
}

/**
 * Generate MCP server configuration for a specific tool
 */
export function generateMcpConfig(
  projectPath: string,
  tool: McpToolKey
): McpServerConfig {
  const config = MCP_TOOL_CONFIGS[tool];

  // Use ${PWD} for tools that support it, otherwise use absolute path
  const projectArg = config.usesProjectVariable
    ? '${workspaceFolder}'
    : projectPath;

  return {
    command: 'npx',
    args: ['-y', '@leanspec/mcp', '--project', projectArg],
  };
}

/**
 * Generate a full MCP config file object
 */
export function generateMcpConfigFile(
  projectPath: string,
  tool: McpToolKey
): McpConfigFile {
  return {
    mcpServers: {
      'lean-spec': generateMcpConfig(projectPath, tool),
    },
  };
}

/**
 * Result of MCP config creation
 */
export interface McpConfigResult {
  tool: McpToolKey;
  configPath: string;
  created?: boolean;
  merged?: boolean;
  skipped?: boolean;
  backupPath?: string;
  error?: string;
}

/**
 * Read existing MCP config file
 */
async function readMcpConfig(configPath: string): Promise<McpConfigFile | null> {
  try {
    const content = await fs.readFile(configPath, 'utf-8');
    return JSON.parse(content) as McpConfigFile;
  } catch {
    return null;
  }
}

/**
 * Write MCP config file with proper formatting
 */
async function writeMcpConfig(
  configPath: string,
  config: McpConfigFile
): Promise<void> {
  const dir = path.dirname(configPath);
  await fs.mkdir(dir, { recursive: true });
  await fs.writeFile(configPath, JSON.stringify(config, null, 2) + '\n', 'utf-8');
}

/**
 * Create or update MCP config for a specific tool
 */
export async function createMcpConfig(
  cwd: string,
  tool: McpToolKey
): Promise<McpConfigResult> {
  const toolConfig = MCP_TOOL_CONFIGS[tool];
  const configPath = path.join(cwd, toolConfig.configPath);
  const absoluteProjectPath = path.resolve(cwd);

  try {
    // Check if config already exists
    const existingConfig = await readMcpConfig(configPath);

    if (existingConfig) {
      // Check if lean-spec is already configured
      if (existingConfig.mcpServers?.['lean-spec']) {
        return {
          tool,
          configPath: toolConfig.configPath,
          skipped: true,
        };
      }

      // Merge: add lean-spec to existing config
      const newServerConfig = generateMcpConfig(absoluteProjectPath, tool);
      existingConfig.mcpServers = existingConfig.mcpServers || {};
      existingConfig.mcpServers['lean-spec'] = newServerConfig;

      await writeMcpConfig(configPath, existingConfig);

      return {
        tool,
        configPath: toolConfig.configPath,
        merged: true,
      };
    }

    // Create new config
    const newConfig = generateMcpConfigFile(absoluteProjectPath, tool);
    await writeMcpConfig(configPath, newConfig);

    return {
      tool,
      configPath: toolConfig.configPath,
      created: true,
    };
  } catch (error) {
    return {
      tool,
      configPath: toolConfig.configPath,
      error: error instanceof Error ? error.message : 'Unknown error',
    };
  }
}

/**
 * Create MCP configs for multiple tools
 */
export async function createMcpConfigs(
  cwd: string,
  tools: McpToolKey[]
): Promise<McpConfigResult[]> {
  const results: McpConfigResult[] = [];

  for (const tool of tools) {
    const result = await createMcpConfig(cwd, tool);
    results.push(result);
  }

  return results;
}

/**
 * Print MCP config results with appropriate formatting
 */
export function printMcpConfigResults(results: McpConfigResult[]): void {
  for (const result of results) {
    const config = MCP_TOOL_CONFIGS[result.tool];

    if (result.created) {
      console.log(chalk.green(`✓ ${config.name}: Created ${result.configPath}`));
    } else if (result.merged) {
      console.log(chalk.green(`✓ ${config.name}: Added lean-spec to ${result.configPath}`));
    } else if (result.skipped) {
      console.log(chalk.yellow(`⚠ ${config.name}: Already configured in ${result.configPath}`));
    } else if (result.error) {
      console.log(chalk.red(`✗ ${config.name}: ${result.error}`));
    }
  }
}

/**
 * Get default MCP tool selection based on detection
 */
export async function getDefaultMcpToolSelection(
  cwd: string
): Promise<{ defaults: McpToolKey[]; detected: McpToolDetectionResult[] }> {
  const detectionResults = await detectMcpTools(cwd);
  const detectedTools = detectionResults
    .filter((r) => r.detected)
    .map((r) => r.tool);

  // If any tools detected, use those as defaults
  if (detectedTools.length > 0) {
    return { defaults: detectedTools, detected: detectionResults };
  }

  // Default to Claude if nothing detected (most common for LeanSpec users)
  return { defaults: ['claude'], detected: detectionResults };
}

/**
 * Parse --mcp-config flag value
 */
export function parseMcpConfigFlag(value: string): McpToolKey[] | 'all' | 'none' {
  if (value === 'all') {
    return 'all';
  }
  if (value === 'none') {
    return 'none';
  }

  const tools = value.split(',').map((t) => t.trim().toLowerCase());
  const validTools: McpToolKey[] = [];

  for (const tool of tools) {
    // Handle aliases
    const normalized = normalizeToolName(tool);
    if (normalized && isValidMcpTool(normalized)) {
      validTools.push(normalized);
    }
  }

  return validTools;
}

/**
 * Normalize tool name aliases
 */
function normalizeToolName(tool: string): McpToolKey | null {
  const aliases: Record<string, McpToolKey> = {
    'claude': 'claude',
    'claude-code': 'claude',
    'claudecode': 'claude',
    'vscode': 'vscode',
    'vs-code': 'vscode',
    'code': 'vscode',
    'copilot': 'vscode',
    'cursor': 'cursor',
    'windsurf': 'windsurf',
  };

  return aliases[tool] || null;
}

/**
 * Check if a tool key is valid
 */
function isValidMcpTool(tool: string): tool is McpToolKey {
  return tool in MCP_TOOL_CONFIGS;
}

File: rust/leanspec-cli/src/commands/init/mcp_config.rs
use serde_json::{json, Map, Value};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
pub enum McpTool {
    Claude,
    VsCode,
    Cursor,
    Windsurf,
}

impl McpTool {
    pub fn name(&self) -> &'static str {
        match self {
            McpTool::Claude => "Claude Code",
            McpTool::VsCode => "VS Code",
            McpTool::Cursor => "Cursor",
            McpTool::Windsurf => "Windsurf",
        }
    }

    pub fn config_path(&self) -> &'static str {
        match self {
            McpTool::Claude => ".mcp.json",
            McpTool::VsCode => ".vscode/mcp.json",
            McpTool::Cursor => ".cursor/mcp.json",
            McpTool::Windsurf => ".windsurf/mcp.json",
        }
    }

    fn uses_project_variable(&self) -> bool {
        matches!(self, McpTool::Claude)
    }

    fn detection_directories(&self) -> &'static [&'static str] {
        match self {
            McpTool::Claude => &[".claude"],
            McpTool::VsCode => &[".vscode"],
            McpTool::Cursor => &[".cursor"],
            McpTool::Windsurf => &[".windsurf"],
        }
    }

    fn detection_files(&self) -> &'static [&'static str] {
        match self {
            McpTool::Claude => &["CLAUDE.md", ".claude.json"],
            McpTool::VsCode => &[],
            McpTool::Cursor => &[".cursorrules"],
            McpTool::Windsurf => &[".windsurfrules"],
        }
    }
}

#[derive(Clone, Debug)]
pub struct McpDetectionResult {
    pub tool: McpTool,
    pub detected: bool,
    pub reasons: Vec<String>,
}

#[derive(Clone, Debug)]
pub struct McpConfigResult {
    pub tool: McpTool,
    pub config_path: PathBuf,
    pub created: bool,
    pub merged: bool,
    pub skipped: bool,
    pub error: Option<String>,
}

pub fn detect_mcp_tools(root: &Path) -> Vec<McpDetectionResult> {
    let mut results = Vec::new();

    for tool in all_tools() {
        let mut reasons: Vec<String> = Vec::new();

        for dir in tool.detection_directories() {
            let dir_path = root.join(dir);
            if dir_path.is_dir() {
                reasons.push(format!("{dir}/ directory found"));
            }
        }

        for file in tool.detection_files() {
            let file_path = root.join(file);
            if file_path.exists() {
                reasons.push(format!("{file} found"));
            }
        }

        results.push(McpDetectionResult {
            tool,
            detected: !reasons.is_empty(),
            reasons,
        });
    }

    results
}

pub fn default_mcp_selection(detections: &[McpDetectionResult]) -> Vec<McpTool> {
    detections
        .iter()
        .filter(|r| r.detected)
        .map(|r| r.tool)
        .collect()
}

pub fn configure_mcp(root: &Path, tools: &[McpTool]) -> Vec<McpConfigResult> {
    let project_path = root.canonicalize().unwrap_or_else(|_| root.to_path_buf());
    let mut results = Vec::new();

    for tool in tools {
        let config_relative = tool.config_path();
        let config_path = root.join(config_relative);

        let result = match write_mcp_config(&config_path, *tool, &project_path) {
            Ok(McpWriteOutcome::Created) => McpConfigResult {
                tool: *tool,
                config_path: PathBuf::from(config_relative),
                created: true,
                merged: false,
                skipped: false,
                error: None,
            },
            Ok(McpWriteOutcome::Merged) => McpConfigResult {
                tool: *tool,
                config_path: PathBuf::from(config_relative),
                created: false,
                merged: true,
                skipped: false,
                error: None,
            },
            Ok(McpWriteOutcome::Skipped) => McpConfigResult {
                tool: *tool,
                config_path: PathBuf::from(config_relative),
                created: false,
                merged: false,
                skipped: true,
                error: None,
            },
            Err(e) => McpConfigResult {
                tool: *tool,
                config_path: PathBuf::from(config_relative),
                created: false,
                merged: false,
                skipped: false,
                error: Some(e.to_string()),
            },
        };

        results.push(result);
    }

    results
}

fn write_mcp_config(
    path: &Path,
    tool: McpTool,
    project_path: &Path,
) -> Result<McpWriteOutcome, Box<dyn std::error::Error>> {
    let parent = path.parent().ok_or("Invalid MCP config path")?;
    fs::create_dir_all(parent)?;

    let existing = read_json(path);
    let mut config = existing
        .clone()
        .unwrap_or_else(|| json!({"mcpServers": {}}));

    let servers_map = match config.get_mut("mcpServers").and_then(Value::as_object_mut) {
        Some(map) => map,
        None => {
            config["mcpServers"] = Value::Object(Map::new());
            config["mcpServers"]
                .as_object_mut()
                .expect("mcpServers object")
        }
    };

    if servers_map.contains_key("lean-spec") {
        return Ok(McpWriteOutcome::Skipped);
    }

    let mcp_entry = build_mcp_entry(tool, project_path);
    servers_map.insert("lean-spec".to_string(), mcp_entry);

    write_json(path, &config)?;

    if existing.is_some() {
        Ok(McpWriteOutcome::Merged)
    } else {
        Ok(McpWriteOutcome::Created)
    }
}

fn build_mcp_entry(tool: McpTool, project_path: &Path) -> Value {
    let project_arg = if tool.uses_project_variable() {
        "${workspaceFolder}".to_string()
    } else {
        project_path.to_string_lossy().to_string()
    };

    json!({
        "command": "npx",
        "args": ["-y", "@leanspec/mcp", "--project", project_arg]
    })
}

fn read_json(path: &Path) -> Option<Value> {
    let content = fs::read_to_string(path).ok()?;
    serde_json::from_str(&content).ok()
}

fn write_json(path: &Path, value: &Value) -> Result<(), Box<dyn std::error::Error>> {
    let content = serde_json::to_string_pretty(value)? + "\n";
    fs::write(path, content)?;
    Ok(())
}

enum McpWriteOutcome {
    Created,
    Merged,
    Skipped,
}

pub fn all_tools() -> Vec<McpTool> {
    vec![
        McpTool::Claude,
        McpTool::VsCode,
        McpTool::Cursor,
        McpTool::Windsurf,
    ]
}
